8229867: Re-examine synchronization usages in http and https protocol handlers

dfuch opened the pull request:
Hi, 

This is a fix that upgrades the old HTTP and HTTPS legacy stack to use virtual-thread friendly locking instead of synchronized monitors.

Most of the changes are mechanical - but there are still a numbers of subtle non-mechanical differences that are outlined below:

1. src/java.base/share/classes/sun/net/www/MessageHeader.java:
   `MessageHeader::print(PrintStream)` => synchronization modified to not synchronize on this while printing
   ( a snapshot of the data is taken before printing instead)

2. src/java.base/share/classes/sun/net/www/MeteredStream.java:
    `MeteredStream::close` was missing synchronization: it is now protected by the lock

3. src/java.base/share/classes/sun/net/www/http/ChunkedOutputStream.java:
    `ChunkedOutputStream` no longer extends `PrintStream` but extends `OutputStream` directly.
    Extending `PrintStream` is problematic for virtual thread. After careful analysis, it appeared that
    `ChunkedOutputStream` didn't really need to extend `PrintStream`. `ChunkedOutputStream`
    is already wrapping a `PrintStream`.  `ChunkedOutputStream` is never returned directly to user
    code but is wrapped in another stream. `ChunkedOutputStream` completely reimplement and
    reverse the flush logic implemented by its old super class`PrintStream` which leads me to believe
    there was no real reason for it to extend `PrintStream` - except for being able to call its `checkError`
    method - which can be done by using `instanceof ChunkedOutputStream` in the caller instead of
    casting to PrintStream. 

4. src/java.base/share/classes/sun/net/www/http/HttpClient.java:
    Synchronization removed from `HttpClient::privilegedOpenServer` and replaced by an `assert`.
    There is no need for a synchronized here as the method is only called from a method that already
    holds the lock.

5. src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java:
    Synchronization removed from `KeepAliveCache::removeVector` and replaced by an `assert`.
    This method is only called from methods already protected by the lock.
    Also the method has been made private.

6. src/java.base/share/classes/sun/net/www/http/KeepAliveStream.java
    `queuedForCleanup` is made volatile as it is read and written directly from outside the class
    and without protection by the `KeepAliveCleanerEntry`.
    Lock protection is also added to `close()`, which was missing it.
    Some methods that have no lock protection and did not need it because only called from
    within code blocks protected by the lock have aquired an `assert isLockHeldByCurrentThread();`

7. Concrete subclasses of `AuthenticationInfo` that provide an implementation for
    `AuthenticationInfo::setHeaders(HttpURLConnection conn, HeaderParser p, String raw)` have
    acquired an `assert conn.isLockheldByCurrentThread();` as the method should only be called
    from within a lock-protected block in `s.n.w.p.h.HttpURLConnection`

8. src/java.base/share/classes/sun/net/www/protocol/http/HttpURLConnection.java
    Several methods in this class have a acquired an `assert isLockheldByCurrentThread();`
    to help track the fact that AuthenticationInfo::setHeaders is only called while
    holding the lock.
    Synchronization was also removed from some method that didn't need it because only
    called from within code blocks protected by the lock:
    `getOutputStream0()`: synchronization removed and replaced by an `assert isLockheldByCurrentThread();`
    `setCookieHeader()`:  synchronization removed and replaced by an `assert isLockheldByCurrentThread();`
    `getInputStream0()`:  synchronization removed and replaced by an `assert isLockheldByCurrentThread();`
    `StreamingOutputStream`: small change to accomodate point 3. above (changes in ChunkedOutputStream).

9. src/java.base/share/classes/sun/net/www/protocol/http/NegotiateAuthentication.java.:
    synchronization removed from `setHeaders` and replace by an assert. See point 7. above.

10. src/java.base/unix/classes/sun/net/www/protocol/http/ntlm/NTLMAuthentication.java:
    synchronization removed from `setHeaders` and replace by an assert. See point 7. above.

11. src/java.base/windows/classes/sun/net/www/protocol/http/ntlm/NTLMAuthentication.java:
    synchronization removed from `setHeaders` and replace by an assert. See point 7. above.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8229867](https://bugs.openjdk.java.net/browse/JDK-8229867): Re-examine synchronization usages in http and https protocol handlers


### Reviewers
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)
 * [Michael McMahon](https://openjdk.java.net/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/558/head:pull/558`
`$ git checkout pull/558`

dfuch commented:
/label security

AlanBateman commented:
Mostly looks good.

HttpClient.openServer - is there a reason to do the SM permission check while holding the lock?

The "closed" field in MeteredStream and ChunkedInputStream needs to be volatile if you really want to read it without holding the lock.

There are a couple of places with comments that I assume you added for yourself when auditing this code. There's one in HttpCapture, and 3 in HttpURLConnection.

dfuch commented:
Thanks Alan!

> HttpClient.openServer - is there a reason to do the SM permission check while holding the lock?

Mostly that was a mechanical change since openServer was synchronized before.
But I guess it seems also desirable for accessing host & port which are protected and not final; 

> The "closed" field in MeteredStream and ChunkedInputStream needs to be volatile if you really want to read it without holding the lock.

I am not so sure. `closed` starts at `false`, and can only moved from `false` to `true`. It can never go back to `false` after it's been set to `true`; So if you observe `true` outside of the lock it does means that it is really closed, isn't it? 
If `closed` is not volatile, the worse that can happen is that you will observe `false` while it's really `true`, and then you will need to pay the price of locking, after which you should be able to see the real `true` value.

> There are a couple of places with comments that I assume you added for yourself when auditing this code. There's one in HttpCapture, and 3 in HttpURLConnection.

Yes - well - I thought that would be mostly beneficial for someone searching for `synchronized` in the java/net and sun/net packages - so that they can determine that the `synchronized` in those places was considered safe and intentionally kept unchanged, and not just overlooked.
I can remove them or reformulate if you think it's better - but I was actually intending to keep these comments.

AlanBateman commented:
> Mostly that was a mechanical change since openServer was synchronized before.
> But I guess it seems also desirable for accessing host & port which are protected and not final;

Okay, I mis-read the old code. The syncrhonization to access host/port isn't obvious so leaving it as is is okay.
 
> I am not so sure. `closed` starts at `false`, and can only moved from `false` to `true`. It can never go back to `false` after it's been set to `true`; So if you observe `true` outside of the lock it does means that it is really closed, isn't it?
> If `closed` is not volatile, the worse that can happen is that you will observe `false` while it's really `true`, and then you will need to pay the price of locking, after which you should be able to see the real `true` value.

Allowing for visibility failures here is confusing for maintainers. If you really want to access closed without the lock (and I don't see any reason to do that) then it would be clearer to all if it were volatile.


> Yes - well - I thought that would be mostly beneficial for someone searching for `synchronized` in the java/net and sun/net packages - so that they can determine that the `synchronized` in those places was considered safe and intentionally kept unchanged, and not just overlooked.
> I can remove them or reformulate if you think it's better - but I was actually intending to keep these comments.

I don't have. a strong opinion here but they did initially look like left over comments. Will they mean anything to someone looking at this code in 2025?

ChrisHegarty reviewed (commented):
Mostly looks good. Just a few comments.

dfuch commented on src/java.base/share/classes/sun/net/www/MeteredStream.java:
```diff
@@ -113,94 +116,140 @@ private boolean isMarked() {
+        if (closed) return -1;
+        lock();
+        try {
+            if (closed) return -1;
```
We could.

dfuch commented on src/java.base/share/classes/sun/net/www/http/KeepAliveStream.java:
```diff
@@ -64,15 +65,9 @@ public KeepAliveStream(InputStream is, ProgressSource pi, long expected, HttpCli
-            return;
-        }
+        // If the inputstream is closed already, or if this stream
+        // has already been queued for cleanup.just return.
```
done

dfuch commented on src/java.base/share/classes/sun/net/www/MessageHeader.java:
```diff
@@ -311,12 +311,23 @@ private boolean isRequestline(String line) {
         at the end. Omits pairs with a null key. Omits
         colon if key-value pair is the requestline. */
-    public synchronized void print(PrintStream p) {
+    private  void print(int nkeys, String[] keys, String[] values, PrintStream p) {
```
Done.

dfuch commented:
From Alan:

> Allowing for visibility failures here is confusing for maintainers. If you really want to access closed without the lock (and I don't see any reason to do that) then it would be clearer to all if it were volatile.

From Chris:

> This double check of closed is kind of irritating. Is it really need, or maybe we just drop it and lock unconditionally?

=> I have removed the double locking.

From Alan:

> I don't have a strong opinion here but they did initially look like left over comments. Will they mean anything to someone looking at this code in 2025?

I have updated the comments to be more explanatory for future maintainer (especially if they use the Annotate feature of the IDE to correlate with the fix in which they were introduced). That said, if you prefer removing them altogether let me know, and I'll remove them.

AlanBateman commented on src/java.base/share/classes/sun/net/www/http/HttpCapture.java:
```diff
@@ -54,6 +54,9 @@
 public class HttpCapture {
+    // Although accessing files could result in blocking operations,
+    // HttpCapture is a corner case; there seem no urgent need to convert
+    // this class to using java.util.concurrent.locks at this time.
```
The updated patch looks good but I think this comment needs another iteration to ensure that it doesn't confuse future maintainers. You could drop it or else replace it with something simple that says that HttpCapture does blocking I/O operations while holding monitors but it's not a concern because it rarely used.

dfuch replied:
Updated as requested.

ChrisHegarty reviewed (approved):
LGTM.
